# DATA CLEANING & PREPARATION
- This notebook will be used to clean the data that will be used for this project.
- The data was obtained from ***https://ckan.africadatahub.org/dataset/kenya-femicide-data***
- In this notebook, we will clean standardize and normalize the data to prepare it for analysis in the EDA notebook.

In [1]:
# Importing libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_excel("../Data/raw/kenya-femicide-data_2016_dec2023.xlsx")
df.head()

,url,title,author_name,medium,country_name,text,published_date,type of murder,name of victim,Age,...,Type of femicide,Murder Scene,Mode of killing,Circumstance,Status on article date,Court date (first appearance),verdict date,Verdict,Years of sentence,Days to verdict
0,http://www.standardmedia.co.ke/article/2000218...,Meru man who killed drunk wife sentenced to hang,Lydiah Nyawira,Standard Digital (Kenya),Kenya,"On August 12, 2005 at around 7pm, Joyce Gacher...",2016-10-05,Femicide,Dorcas Kaguri,Unknown,...,Intimate,home,blunt force,Argument,Unknown,2005-08-13 00:00:00,2016-06-24 00:00:00,Guilty,Death,4061
1,http://www.standardmedia.co.ke/article/2001300...,"Court: Kakamega man guilty of raping, killing ...",Jack Murima,Standard Digital (Kenya),Kenya,"Jack Murima - Sun, 28. October 2018 12:54 PM -...",2018-10-29,Femicide,Selina Ikambi Anasi,79,...,Non-intimate,Public Space,hacked,Unknown,Unknown,Unknown,2018-10-28 00:00:00,guilty,awaiting ruling,2817
2,http://www.standardmedia.co.ke/article/2001241...,My wife 'died from fall in fight over volume o...,Kamau Muthoni,Standard Digital (Kenya),Kenya,Former journalist Moses Dola Otieno yesterday ...,2017-05-25,Femicide,Wambui Kabiru,Unknown,...,Intimate,home,blunt force,Argument,Unknown,Unknown,2018-10-05 00:00:00,guilty,10,2714
3,http://www.nation.co.ke/news/Kabogo-cleared-Me...,Kabogo cleared in varsity student's murder case,ELISHA OTIENO,Daily Nation (Kenya),Kenya,Kiambu Governor Kabogo and five others adverse...,2016-10-07,Femicide,Mercy Keino,Unknown,...,Unknown,Public Space,Unknown,Unknown,Unknown,Unknown,2016-10-07 00:00:00,Not Guilty,Freed,1939
4,http://www.mediamaxnetwork.co.ke/news/pastor-t...,"Pastor to serve 25 years in jail for defiling,...",People Daily,MediaMax Network,Kenya,Musa Wekesa A pastor from Kitale has been sent...,2019-05-25,Femicide,Scholastica Mmbihi,Unknown,...,Unknown,home,stabbed,Argument,Unknown,Unknown,2019-05-25 00:00:00,guilty,25,2879


In [3]:
df.shape

(507, 24)

In [4]:
df.columns

Index(['url', 'title', 'author_name', 'medium', 'country_name', 'text',
       'published_date', 'type of murder', 'name of victim', 'Age',
       'date of murder', 'Location', 'name of suspect', 'suspect relationship',
       'Type of femicide', 'Murder Scene', 'Mode of killing', 'Circumstance',
       'Status on article date', 'Court date (first appearance)',
       'verdict date', 'Verdict', 'Years of sentence', 'Days to verdict'],
      dtype='object')

In [37]:
df.describe()

,url,title,author_name,medium,country_name,text,published_date,type of murder,name of victim,Age,...,Type of femicide,Murder Scene,Mode of killing,Circumstance,Status on article date,Court date (first appearance),verdict date,Verdict,Years of sentence,Days to verdict
count,507,507,507,507,507,507,507,507,507,507,...,507,507,507,507,507,507,507,507,507,507
unique,507,477,285,26,1,469,427,2,399,65,...,7,10,13,48,15,67,37,6,20,37
top,http://www.standardmedia.co.ke/article/2000218...,Femicide,Unknown,The Star (Kenya),Kenya,Femicide was committed,Unknown,Femicide,Unnamed,Unknown,...,Intimate,home,stabbed,Unknown,under investigation,Unknown,Unknown,Unknown,Unknown,#VALUE!
freq,1,28,43,188,507,39,28,493,103,195,...,311,256,143,218,274,440,471,469,471,471


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 507 entries, 0 to 506
Data columns (total 24 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   url                            507 non-null    object        
 1   title                          479 non-null    object        
 2   author_name                    473 non-null    object        
 3   medium                         495 non-null    object        
 4   country_name                   492 non-null    object        
 5   text                           468 non-null    object        
 6   published_date                 479 non-null    datetime64[ns]
 7   type of murder                 507 non-null    object        
 8   name of victim                 507 non-null    object        
 9   Age                            507 non-null    object        
 10  date of murder                 507 non-null    object        
 11  Location           

In [7]:
# Check the number of missing values in each column
for col in df.columns:
    print(f'{col}: {df[col].isna().sum()}')

url: 0
title: 28
author_name: 34
medium: 12
country_name: 15
text: 39
published_date: 28
type of murder: 0
name of victim: 0
Age: 0
date of murder: 0
Location: 2
name of suspect: 39
suspect relationship: 2
Type of femicide: 59
Murder Scene: 0
Mode of killing: 0
Circumstance: 69
Status on article date: 71
Court date (first appearance): 99
verdict date: 0
Verdict: 101
Years of sentence: 102
Days to verdict: 0


### Filling Na VAlues column by column
Since we only 507 records rather than droping the Na values its better to fill them while keeping the integrity of the data

In [10]:
# Filling Title column with Femicide
df['title'] = df['title'].fillna('Femicide')

#Checking if there is still missing values in that column
df['title'].isna().sum()

np.int64(0)

In [14]:
# Filling authour name with unknown
df['author_name'] = df['author_name'].fillna('Unknown')
df['author_name'].isna().sum()

np.int64(0)

In [ ]:
# Filling Medium with the most frequent value
df['medium'] = df['medium'].fillna(df['medium'].mode()[0])
df['medium'].isna().sum()

np.int64(0)

In [ ]:
#Filling country_name with kenya
df['country_name'] = df['country_name'].fillna('Kenya')
df['country_name'].isna().sum()

np.int64(0)

In [23]:
df['text'] = df['text'].fillna('Femicide was committed')
df['text'].isna().sum()

np.int64(0)

In [26]:
df['published_date'] =  pd.to_datetime(df['published_date'])
df['published_date'][299]

Timestamp('2020-09-15 00:00:00')

In [27]:
#Filling missing published dates with Unkown
df['published_date'] = df['published_date'].fillna('Unknown')
df['published_date'].isna().sum()

np.int64(0)

In [30]:
# Location with Uknown
df['Location'] = df['Location'].fillna('Unknown')
df['Location'].isna().sum()

np.int64(0)

In [31]:
df['name of suspect'] = df['name of suspect'].fillna('Unknown')
df['name of suspect'].isna().sum()

np.int64(0)

In [32]:
df['suspect relationship'] = df['suspect relationship'].fillna('Unknown')
df['suspect relationship'].isna().sum()

np.int64(0)

In [33]:
#Type of femicide
df['Type of femicide'] = df['Type of femicide'].fillna('Unknown')
df['Type of femicide'].isna().sum()

np.int64(0)

In [35]:
df[['Circumstance', 'Status on article date', 'Court date (first appearance)', 'verdict date', 'Verdict',
'Years of sentence', 'Days to verdict']] = df[['Circumstance', 'Status on article date', 'Court date (first appearance)', 'verdict date', 'Verdict',
'Years of sentence', 'Days to verdict']].fillna('Unknown')

#Checking the whole dataset for mising values in each column
for col in df.columns:
    print(f'{col}: {df[col].isna().sum()}')

url: 0
title: 0
author_name: 0
medium: 0
country_name: 0
text: 0
published_date: 0
type of murder: 0
name of victim: 0
Age: 0
date of murder: 0
Location: 0
name of suspect: 0
suspect relationship: 0
Type of femicide: 0
Murder Scene: 0
Mode of killing: 0
Circumstance: 0
Status on article date: 0
Court date (first appearance): 0
verdict date: 0
Verdict: 0
Years of sentence: 0
Days to verdict: 0


- Once Null values have been filled we can now proceed to checking the data types  of each column and validate them
- I find the best way for this is to use value_counts() method to see if the data types do match the expected values

In [40]:
# Check the top 5 values of each column
for col in df.columns:
    print(f'{col}:\n{df[col].value_counts().head()}')

url:
url
http://www.standardmedia.co.ke/article/2000218467/meru-man-who-killed-drunk-wife-sentenced-to-hang               1
http://www.standardmedia.co.ke/article/2001300764/court-kakamega-man-guilty-of-raping-killing-his-grandmother    1
http://www.standardmedia.co.ke/article/2001241021/my-wife-died-from-fall-in-fight-over-volume-of-radio           1
http://www.nation.co.ke/news/Kabogo-cleared-Mercy-Keino-murder-inquest/1056-3408370-1p17r1z/index.html           1
http://www.mediamaxnetwork.co.ke/news/pastor-to-serve-25-years-in-jail-for-defiling-killing-student-527203       1
Name: count, dtype: int64
title:
title
Femicide                                                                28
Muchatha killings puzzle as ruthless gang rape, kills, maims at will     2
Bungoma man murders, buries wife in shallow grave behind their house     2
Kisumu man arrested for killing wife after domestic quarrel              2
Pastor to serve 25 years in jail for defiling, killing student           1
Na